lensless/rfft_convolve.py

```diff
@@ -68,50 +68,68 @@ def __init__(self, psf, dtype=None, pad=True, norm="ortho", **kwargs):
                 self._pad(self._psf), norm=norm, dim=(0, 1), s=self._padded_shape[:2]
             )
             self._Hadj = torch.conj(self._H)
-            self._padded_data = torch.zeros(size=self._padded_shape, dtype=dtype, device=psf.device)
-
+            # self._padded_data = torch.zeros(size=self._padded_shape, dtype=dtype, device=psf.device)
         else:
             self._H = fft.rfft2(self._pad(self._psf), axes=(0, 1), norm=norm)
             self._Hadj = np.conj(self._H)
             self._padded_data = np.zeros(self._padded_shape).astype(dtype)
 
+        self.dtype = dtype
+
     def _crop(self, x):
-        return x[self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]]
+        if len(x.shape) <= 3:  # Only a single image
+            return x[self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]]
+        elif len(x.shape) == 4:  # we are processing a batch
+            return x[
+                :, self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1], :
+            ]
 
     def _pad(self, v):
-        if self.is_torch:
-            vpad = torch.zeros(size=self._padded_shape, dtype=v.dtype, device=v.device)
-        else:
-            vpad = np.zeros(self._padded_shape).astype(v.dtype)
-        vpad[self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]] = v
-        return vpad
+        if len(v.shape) <= 3:  # Only a single image
+            if self.is_torch:
+                vpad = torch.zeros(size=self._padded_shape, dtype=v.dtype, device=v.device)
+            else:
+                vpad = np.zeros(self._padded_shape).astype(v.dtype)
+            vpad[self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]] = v
+            return vpad
+        elif len(v.shape) == 4:  # we are processing a batch
+            batch_size = v.shape[0]
+            if self.is_torch:
+                vpad = torch.zeros(
+                    size=[batch_size] + self._padded_shape, dtype=v.dtype, device=v.device
+                )
+            else:
+                raise ValueError("Batch processing only allowed on torch tensor")
+            vpad[
+                :, self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]
+            ] = v
+            return vpad
 
     def convolve(self, x):
         """
         Convolve with pre-computed FFT of provided PSF.
         """
-
         if self.pad:
-            self._padded_data[
-                self._start_idx[0] : self._end_idx[0], self._start_idx[1] : self._end_idx[1]
-            ] = x
+            self._padded_data = self._pad(x)
         else:
-            self._padded_data[:] = x
+            if self.is_torch:
```

Comment:
keep original? as same for torch and numpy, and if `self._padded_data` is defined in `init` can used `[:]`

Reply:
And I think it's best practice to define object variables in the `init`